exonum/src/messages2/protocol.rs

```diff
@@ -0,0 +1,660 @@
+            use $crate::events::error::into_failure;
+                match message.message_class() {
+                    $($cls_num =>
+                        match message.message_type() {
+                            $($typ_num =>{
+                                let payload = $typ::deserialize(message.payload())
+                                                .map_err(into_failure)?;
+                                let message = Message::new(payload, message);
+                                Ok($proto_name::$cls($cls::$typ(message)))
+                            }),+
+                            _ => bail!("Not found message with this type {}", message.message_type())
+                        }
+                    ),+
+                    _ => bail!("Not found message with this class {}", message.message_class())
+                }
+            }
+
+            pub fn signed_message(&self) -> &SignedMessage {
+                match *self {
+                    $(
+                        $proto_name::$cls(ref c) => {
+                            match *c {
+                                $(
+                                    $cls::$typ(ref t) => {
+                                        t.signed_message()
+                                    }
+                                ),+
+                            }
+                        }
+                    ),+
+                }
+            }
+        }
+    };
+}
+
+impl_protocol!{
+    SignedMessage => Protocol {
+        0 => Service {
+            RawTransaction = 0,
+            Connect = 1,
+            Status = 2,
+        },
+        1 => Consensus {
+            Precommit = 0,
+            Propose = 1,
+            Prevote = 2,
+        },
+        3 => Responses {
+            TransactionsResponse = 0,
+            BlockResponse = 1
+        },
+        3 => Requests {
+            ProposeRequest = 0,
+            TransactionsRequest = 1,
+            PrevotesRequest = 2,
+            PeersRequest = 3,
+            BlockRequest = 4,
+        },
+
+    }
+}
+
+
+impl Protocol {
+
+    /// Creates new protocol message.
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
+    pub fn new<T: ProtocolMessage>(message: T, author: PublicKey, secret_key: &SecretKey)
+                                -> Protocol {
+        let signed = SignedMessage::new(message, author, secret_key);
+        Self::deserialize(signed).expect("Couldn't deserialize newly created message.")
+    }
+
+    /// Creates new protocol message.
+    /// Return concrete `Message<T>`
+    ///
+    /// # Panics
+    ///
+    /// On serialization fail this method can panic.
+    pub fn concrete<T: ProtocolMessage>(message: T, author: PublicKey, secret_key: &SecretKey)
+                                   -> Message<T> {
+        T::try_from(Self::new(message, author, secret_key))
+            .expect("BUG: Newly created message matched not as transaction.")
+    }
```

Comment:
These back-and-forth conversions in `new()` and `concrete()` with expected panics bother me. Is there a better way to do this, or this is just quick-and-dirty implementation?